### ISSUE TITLE
Support nullable return types in ExecuteScalar

### ISSUE THREAD

**Issue by HeinziAT** [praeclarum/sqlite-net#450]
Currently, executing something like

    var myValue = myDb.ExecuteScalar<int?>("SELECT myNullableField FROM ...");

yields an error: `System.NotSupportedException: Don't know how to read System.Nullable``1[System.Int32]`.

I think that this can be fixed quite easily by changing the following lines in `SQLiteCommand.ExecuteScalar<T>` from

    val = (T)ReadCol (stmt, 0, colType, typeof(T));

to

    var underlyingClrType = Nullable.GetUnderlyingType(typeof(T)) ?? typeof(T);
    val = (T)ReadCol (stmt, 0, colType, underlyingClrType);

**Comment by rwil02** [comment]
Hmm.  I'd think that if ReadCol then returns DBNull you'd get a cast failure

**Comment by HeinziAT** [comment]
ReadCol won't return DBNull, see the first two lines of the method.

**Comment by rwil02** [comment]
That depends on SQLite3.ColumnType (stmt, i)  returning SQLite3.ColType.Null
Which is a resultant of calling sqlite3_column_type, which does indeed vary every time it's called.
The ways it's named I would have expected it to return (for instance) 'Int32' as per DataColumn.DataType or IDataReader.GetFieldType 
This is more like GetValue().GetType()
Live and learn.
But shouldn't it then become sensible to have `ExecuteScalar<string>` and `T? ExecuteScalar<T>() where T : struct`?
Oh, but then you wouldn't get cast failures if you wanted them.

**Comment by thomaslevesque** [comment]
The problem here is that there would be no way to distinguish between "no result" and "null value". In fact, the problem already exists now : `ExecuteScalar<int>` will return 0 if there is no matching row.

> But shouldn't it then become sensible to have `ExecuteScalar<string>` and `T? ExecuteScalar<T>()` where T : struct?

You mean this?
```
public T ExecuteScalar<T> (string query, params object[] args) where T : class;
public T? ExecuteScalar<T> (string query, params object[] args) where T : struct
```

Unfortunately it wouldn't work; you can't have two overloads whose signatures differ only by generic constraint...

**Comment by HeinziAT** [comment]
> The problem here is that there would be no way to distinguish between "no result" and "null value". In fact, the problem already exists now : ExecuteScalar<int> will return 0 if there is no matching row.

Indeed. My proposed change would actually improve this situation for non-nullable database fields, since `ExecuteScalar<int?>` would return `null` in this case (and "0" would really mean "0"). For nullable database fields, the ambiguity between "no rows" and "null value" remains.

For what it's worth: Some time ago, I needed to switch to a fork of this project for technical reasons. They [already implemented](https://github.com/oysteinkrog/SQLite.Net-PCL/blob/master/src/SQLite.Net/SQLiteCommand.cs) something similar to my proposed change, and my experience has been very positive.
